test/jdk/java/lang/annotation/LoaderLeakTest.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2004, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 5040740
+ * @summary annotations cause memory leak
+ * @library /test/lib
+ * @build jdk.test.lib.process.*
+ * @run testng LoaderLeakTest
+*/
+
+import jdk.test.lib.Utils;
+import jdk.test.lib.process.ProcessTools;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import java.io.FileInputStream;
+import java.lang.annotation.Retention;
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.nio.file.*;
+import java.util.*;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+public class LoaderLeakTest {
+
+    @Test
+    public void testWithoutReadingAnnotations() throws Throwable {
+        runJavaProcessExpectSuccessExitCode("Main");
+    }
+
+    @Test
+    public void testWithReadingAnnotations() throws Throwable {
+        runJavaProcessExpectSuccessExitCode("Main",  "foo");
+    }
+
+    private void runJavaProcessExpectSuccessExitCode(String ... command) throws Throwable {
+        var processBuilder = ProcessTools.createJavaProcessBuilder(command)
+                                                      .directory(Paths.get(Utils.TEST_CLASSES).toFile());
+        ProcessTools.executeCommand(processBuilder).shouldHaveExitValue(0);
+    }
+
+}
+
+class Main {
+
+    public static void main(String[] args) throws Exception {
+        for (int i = 0; i < 100; i++) {
+            doTest(args.length != 0);
+        }
+    }
+
+    static void doTest(boolean readAnn) throws Exception {
+        ClassLoader loader = new SimpleClassLoader();
+        var c = new WeakReference<Class<?>>(loader.loadClass("C"));
+        if (c.get() == null) throw new AssertionError("class missing after loadClass");
+        if (c.get().getClassLoader() != loader) throw new AssertionError("wrong classloader");
```

Comment:
This can throw NPE as there is no guarantee that two subsequent calls will return the same value: c.get() could very well be null the second time around - although yes - arguably, `c.get()` should not return `null` here since there should still be a strong reference to `loader`.